### ISSUE TITLE
Kernel: Kernel Address Sanitizer (KASAN) Integration

### ISSUE THREAD

**Issue by bgianfo** [SerenityOS/serenity#5351]
This issue is meant to track progress on the implementation of KASAN in SerenityOS. 

### Description
KASAN is a dynamic analysis tool that finds memory errors. It focuses
mostly on finding use-after-free and out-of-bound read/writes bugs.

KASAN works by having the kernel allocate a "shadow memory" region which is
used to store whether each byte of memory is safe to access. The compiler then
instruments the kernel code and a check is inserted which validates the state of the
shadow memory region on every memory access (load or store). The check then
calls back into the kernel, allowing it to validate the shadow memory state. Kernel
memory allocators are then enlightened to record the state of allocations/frees in
the shadow memory region as they occur. 

The student who worked on KASAN for NetBSD during Google SOC has a very nice diagram in [this blog article](https://blog.netbsd.org/tnf/entry/kernel_address_sanitizer_part_3): 

![image](https://user-images.githubusercontent.com/1212/107890347-72232d80-6f10-11eb-973b-78782fe1d532.png)

### References:

 ASAN Paper: https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/37752.pdf
 KASAN Docs: https://github.com/google/kasan
 NetBSD KASAN Blog: https://blog.netbsd.org/tnf/entry/kernel_address_sanitizer_part_3
 LWN KASAN Article: https://lwn.net/Articles/612153/

### Execution Plan

To fully integrate KASAN into the SerenityOS kernel we need to:

- [X] a) Implement the KASAN interface to intercept the injected loads/stores.

    ```
      void __asan_load*(address);
      void __asan_store(address); 
    ```

- [ ] b) Setup KASAN region and determine the shadow memory offset + translation.
    This might be challenging since Serenity is only 32bit at this time.

     Ex: Linux implements kernel address -> shadow address translation like:
    ```
  static inline void *kasan_mem_to_shadow(const void *addr)
  {
      return ((unsigned long)addr >> KASAN_SHADOW_SCALE_SHIFT)
              + KASAN_SHADOW_OFFSET;
  }
    ```

- [ ] c) Integrating KASAN with Kernel allocators.
      The kernel allocators need to be taught how to record allocation state
      in the shadow memory region.

### Progress

- [X] PR #5350 - Kernel: Initial integration of Kernel Address Sanitizer (KASAN) 
      This PR address most of bullet "a" above. Implement the KASAN interface to allow the Kernel to link clean.
      The resulting kernel is not currently usable, but that's ok for now.

**Comment by Stupremee** [comment]
I would like to give this a try 👍

**Comment by bgianfo** [comment]
Sure, I would appreciate any help 😁

**Comment by bgianfo** [comment]
Fixed up the build: #9106
